File: BEmu3/HistoricalDataRequest/EventHistoric.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace BEmu3.HistoricalDataRequest
{
    public class EventHistoric : Event
    {
        private readonly List<Message> _messages;

        internal EventHistoric(RequestHistoric request)
        {
            base._request = request;
            this._messages = this.GenerateMessages();
        }

        public override IEnumerable<Message> GetMessages()
        {
            return this._messages;
        }

        private List<Message> GenerateMessages()
        {
            List<Message> result = new List<Message>();
            var hreq = (RequestHistoric)base._request;
            Random rand = new Random(64);
            
            foreach (var security in hreq.Securities)
            {
                var fieldData = new Dictionary<DateTime, Dictionary<string, object>>();

                if (hreq.DtStart.HasValue)
                {
                    foreach (var dtCurrent in hreq.GetDates())
                    {
                        if ((dtCurrent.DayOfWeek != DayOfWeek.Sunday) && (dtCurrent.DayOfWeek != DayOfWeek.Saturday))
                        {
                            var fields = new Dictionary<string, object>();
                            fieldData.Add(dtCurrent, fields);

                            foreach (var field in hreq.Fields)
                            {
                                fields.Add(field, Math.Round(rand.NextDouble() * 100d, 2));
                            }
                        }
                    }
                }

                MessageHistoric msg = new MessageHistoric(base._request.correlationId, security, fieldData, result.Count);
                result.Add(msg);
            }

            return result;
        }

    }
}

File: BEmuJava/src/BEmu/HistoricalDataRequest/EventHistoric.java
package BEmu.HistoricalDataRequest;

import java.util.Calendar;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.LinkedHashMap; //preserves insertion order
import BEmu.Event;
import BEmu.Message;
import BEmu.Datetime;

public class EventHistoric extends Event
{
    public EventHistoric(RequestHistoric request)
    {
        super._request = request;
        super._messages = this.generateMessages();
    }

    private List<Message> generateMessages()
    {
    	List<Message> result = new ArrayList<Message>();
    	RequestHistoric hreq = (RequestHistoric)super._request;
    	
    	List<String> securities = hreq.securities();
    	for(int i = 0; i < securities.size(); i++)
    	{
    		String security = securities.get(i);
    		Map<Datetime, Map<String, Object>> fieldData = new LinkedHashMap<Datetime, Map<String, Object>>();
    		
    		if(hreq.dtStart() != null)
    		{
    			List<Datetime> dates = hreq.getDates();
    			for(int j = 0; j < dates.size(); j++)
    			{
    				Datetime current = dates.get(j);
    				if((current.calendar().get(Calendar.DAY_OF_WEEK) != Calendar.SUNDAY) && (current.calendar().get(Calendar.DAY_OF_WEEK) != Calendar.SATURDAY))
    				{
    					Map<String, Object> fields = new LinkedHashMap<String, Object>();
    					
    					List<String> hfields = hreq.fields();
    					for(int k = 0; k < hfields.size(); k++)
    					{
    						fields.put(hfields.get(k), BEmu.Types.RandomDataGenerator.randomDouble());
    					}

    					fieldData.put(current, fields);
    				}
    			}
    		}    		

            MessageHistoric msg = new MessageHistoric(super._request.correlationId(), security, fieldData, result.size());
            result.add(msg);    		
    	}
    	
    	return result;
    }
    
}
